keps/1282-pods-ready-requeue-strategy/README.md

```diff
@@ -75,25 +107,69 @@ Consider including folks who also work outside the SIG or subproject.
 
 ### API Changes
 
-Add an additional field to the Kueue ConfigMap to allow administrators to specify what timestamp to consider during queue sorting (under the pre-existing waitForPodsReady block).
+#### KueueConfig
+
+Add an additional field to the KueueConfig to allow administrators to specify what timestamp to consider during queue sorting (under the pre-existing waitForPodsReady block).
 
 Possible settings:
 
 * `Eviction` (Back of queue)
 * `Creation` (Front of queue)
 
-```yaml
-kind: ConfigMap
-metadata:
-  name: kueue-manager-config
-  namespace: kueue-system
-data:
-  controller_manager_config.yaml: |
-    apiVersion: config.kueue.x-k8s.io/v1beta1
-    kind: Configuration
-    # ...
-    waitForPodsReady:
-      requeuingTimestamp: Creation | Eviction # <-- New field
+```go
+type WaitForPodsReady struct {
+	...
+	// requeuingStrategy defines the strategy for requeuing a Workload
+	// +optional
+	RequeuingStrategy *RequeuingStrategy `json:"requeuingStrategy,omitempty"`
+}
+
+type RequeuingStrategy struct {
+	// timestamp defines the timestamp used for requeuing a Workload
+	// that was evicted due to Pod readiness. Defaults to Eviction.
+	// +optional
+	Timestamp *RequeuingTimestamp `json:"timestamp,omitempty"`
+	
+	// backOffLimitCount defines the maximum number of requeuing retries.
+	// When the number is reached, the workload is deactivated.	
+	//
+	// Defaults to null. 
+	// +optional
+	BackOffLimitCount *int32 `json:"backOffLimitCount,omitempty"`
+    
+	// backOffLimitTimeout defines the time for a workload that 
+	// has once been admitted to reach the PodsReady=true condition. 
+	// When the time is reached, the workload is deactivated.
+	// 	
+	// Defaults to null.
+	// +optional
+	BackOffLimitTimeout *int32 `json:"backOffLimitTimeout,omitempty"`
+}
+
+type RequeuingTimestamp string
+
+const (
+	// creationTimestamp timestamp (from Workload .metadata.creationTimestamp).
+	CreationTimestamp RequeuingTimestamp = "Creation"
+    
+	// evictionTimestamp timestamp (from Workload .status.conditions).
+	EvictionTimestamp RequeuingTimestamp = "Eviction"
+)
+```
+
+#### Workload
+
+Add a new field, "requeuedCount", to the Workload to allow recording the number of times a workload is requeued.
+
+```go
+type WorkloadStatus struct {
+	...
+	// requeuedCount records the number of times a workload has been requeued.
+	// When a deactivated workload is reactivated, this count is reset to 0. 
+	//
+	// +optional
+	RequeuedCount *int32 `json:"requeuedCount,omitempty"`
```

Comment:
How about adding the last requeue / backoff timestamp and the backoff duration, so it's possible to know when the workload will be "retried", without recomputing the exponential backoff logic for an external observer?

Reply:
+1
I wanted to keep the API as minimal as possible, but I agree that usability can be compromised without a `maxBackoff`

Reply:
SGTM
I will extend this API.

Reply:
can you add it? We are almost closing the release.

Reply:
I should do ASAP...

Reply:
@alculquicondor Here are quick API design sharing: Please let me know if the API design wouldn't be expected.

```golang
type WorkloadStatus struct {
	...
	// requeueState records 
	// 
	// +optional
	RequeueState *RequeueState `json:"requeueState,omitempty"` 
}

type RequeueState struct {
	// count records the number of times a workload has been requeued.
	// When a deactivated workload is reactivated, this count is reset to 0. 
	//
	// +optional
	Count *int32 `json:"requeuedCount,omitempty"`

	// +optional
	LastBackoffTime metav1.Time `json:"lastRequeuedTime,omitempty"`
	
	// +optional
	BackoffDuration metav1.Duration `json:"backoffDuration,omitempty"`
}
```

I'm organizing the above API.

Reply:
instead of `BackoffDuration` and `lastBackoffTime`, it might be better to have just `requeueAt metav1.Time` that indicates the time at which Kueue will consider this workload again for admission.

The time when the workload was evicted is already visible in the condition.

Reply:
> instead of BackoffDuration and lastBackoffTime, it might be better to have just requeueAt metav1.Time that indicates the time at which Kueue will consider this workload again for admission.
> 
> The time when the workload was evicted is already visible in the condition.

That makes sense. Even if we add only `requeueAt`, we can avoid recomputing the exponential backoff logic.
After we truly want to have `duration` based on user feedback, we can extend API.

Reply:
Done.